Public/Src/Engine/Scheduler/PipExecutor.cs

```diff
@@ -934,6 +934,40 @@ private static void MakeSharedOpaqueOutputIfNeeded(IPipExecutionEnvironment envi
             }
         }
 
+        /// <summary>
+        /// Report change affected ouputs to file content manager
+        /// </summary>
+        internal static void ReportSourceChangeAffectedOutputs(
+            IPipExecutionEnvironment environment,
+            PipResultStatus status,
+            ReadOnlyArray<AbsolutePath> dynamicallyObservedFiles,
+            ReadOnlyArray<AbsolutePath> dynamicallyObservedEnumerations,
+            Pip pip,
+            IReadOnlyCollection<(FileArtifact, FileMaterializationInfo, PipOutputOrigin)> outputContent = null,
+            IReadOnlyCollection<DirectoryArtifact> directoryOutputs = null,
+            IReadOnlyCollection<AbsolutePath> sourceChangeAffectedOutputFiles = null,
+            IReadOnlyCollection<AbsolutePath> sourceChangeAffectedOutputDirectroies = null)
+        {
+            PipOutputOrigin? overrideOutputOrigin = null;
+            if (status == PipResultStatus.NotMaterialized)
+            {
+                overrideOutputOrigin = PipOutputOrigin.NotMaterialized;
+            }
+
+            if (pip.IsOutputAffectedBySourceChange(dynamicallyObservedFiles, dynamicallyObservedEnumerations, environment.Context.PathTable, sourceChangeAffectedOutputFiles, sourceChangeAffectedOutputDirectroies))

```

Comment:
>pip.IsOutputAffectedBySourceChange [](start = 16, length = 34)

Why are we doing it here? If you look the `OnPipCompleted` method of `Scheduler`, we update the incremental scheduling state with dynamically observed observations there. Why can't we do it in `OnPipCompleted`?

Reply:
I moved the reporting to the PipExecutionStep.ExecuteProcess and only do the reporting on the master. I can't do that in the OnPipCompleted, since the master would not has OnPipCompleted for all pips. But master has ExecuteProcess for all pips.